File: 문민우/프로그래머스/level2/방문길이.java
package level2;

import java.util.Set;
import java.util.HashSet;

public class 방문길이 {
    public int solution(String dirs) {

        int answer = 0;

        Set<String> set = new HashSet<>();  // 이동 경로
        int nowX = 0;
        int nowY = 0;

        for (int i = 0; i < dirs.length(); i++) {

            int nextX = nowX;
            int nextY = nowY;
            String road = "";

            switch (dirs.charAt(i)) {  // (0, 0) -> (0, 1) & (0, 1) -> (0, 0) => 0001 경로

                case 'U':
                    nextY++;
                    road += nowX;
                    road += nowY;
                    road += nextX;
                    road += nextY;
                    break;

                case 'D':
                    nextY--;
                    road += nextX;
                    road += nextY;
                    road += nowX;
                    road += nowY;
                    break;

                case 'R':
                    nextX++;
                    road += nowX;
                    road += nowY;
                    road += nextX;
                    road += nextY;
                    break;

                case 'L':
                    nextX--;
                    road += nextX;
                    road += nextY;
                    road += nowX;
                    road += nowY;
            }

            if (nextX < -5 || nextY < -5 || nextX > 5 || nextY > 5)
                continue;

            set.add(road);

            // 범위를 벗어나지 않았다면 갱신
            nowX = nextX;
            nowY = nextY;
        }

        answer = set.size();

        return answer;
    }
}

File: 이솔/프로그래머스/Level2/방문길이.py
def solution(dirs):
    x,y=0,0
    answer = 0
    arr=[(0,0,"N")]
    for i in dirs:
        if i == "U":
            if y==5 :
                continue
            if (x,y,"U") not in arr and (x,y+1,"D") not in arr:
                answer+=1
                arr.append((x,y,"U"))
                arr.append((x,y+1,"D"))
            y+=1
        if i =="D":
            if  y==-5:
                continue
            if (x,y,"D") not in arr and (x,y-1,"U") not in arr:
                answer+=1
                arr.append((x,y,"D"))
                arr.append((x,y-1,"U"))
            y-=1
        if i=="R":
            if x==5 :
                continue
            
            if (x,y,"R") not in arr and (x+1,y,"L") not in arr:
                answer+=1
                arr.append((x,y,"R"))
                arr.append((x+1,y,"L"))
            x+=1
        if i=="L":
            if x==-5:
                continue
            
            if (x,y,"L") not in arr and (x-1,y,"R") not in arr:
                answer+=1
                arr.append((x,y,"L"))
                arr.append((x-1,y,"R"))
            x-=1
        
    return answer